### DISCUSSION TITLE
remix-express template use `.mjs` file extension for `server.js` entry point (es-modules)

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [remix-run/remix#5497] [Q&A]
Currently the remix-express template, generates a `server.js` file (common-js format). 
How could I make `purgeRequireCache` work when using converting my server.js to es-modules (server.mjs)?
The `require` & `require.cache` wont work if you convert the file to es-module.

Is there a recommended approach for using es-modules format (`server.js` ---> `server.mjs) while 
![CleanShot 2023-02-17 at 02 44 20](https://user-images.githubusercontent.com/6743796/219623140-f33f6bf2-f1de-475f-b2c4-3fa250dcb975.png)

**Comment by cliffordfajardo** [comment]
@pcattori - After the new remix dev server work is stabilized, I believe this might no longer be a concern, since with the new remix dev server, you won't need to use `purgeRequireCache` at all
https://github.com/remix-run/remix/issues/5240

Mostly decided to write up this Q&A discussion in case future remix users on older versions of remix or using older dev server see this while attempting to use `.mjs` file for their `server.js` file

@machour @MichaelDeBoey - we'll probably need to create a follow up ticket to update, what do you think?
1. all remix examples in the https://github.com/remix-run/examples
2. upgrade the [remix-expres package](https://github.com/remix-run/remix/tree/main/packages/remix-express)
3. Are there any other packages which use a `server.js`, that might need to change to `server.mjs` (or even server.ts)?

I was going to create an issue about this, but didn't want to jump into things without verifying

### References 
 - Remix Roadmap: https://github.com/orgs/remix-run/projects/5/views/1?filterQuery=hmr&pane=issue&itemId=17985756

**Reply by cliffordfajardo** [reply]
This work @mcansh started a few days might be relevant 👀 ?
https://github.com/remix-run/remix/issues/5452

**Reply by pcattori** [reply]
To be clear, the new dev server will make it so that in-memory stuff sticks around _across navigations_, not _across rebuilds_. That would require server-side HMR (require cache purging is coarse-grain version of that). Which is actually fairly easy to implement in CJS, but since there's no API to mess with the `import` cache, I don't think its possible in ESM.

Unless the same "modified timestamp as a query parameter" in `import` works in Node and not just in the browser? Haven't tried that.

**Comment by cliffordfajardo** [comment]
Remix community memeber @xHomu  created a really nice repository showing how to setup the new v2 dev server 🎉 

Github Repo: https://github.com/xHomu/remix-v2-server
I discovered this via the https://remix.guide/ website ❤️ 

↓↓↓
<a target="_blank" href="https://github.com/xHomu/remix-v2-server">
  <img height="400" src="https://github.com/remix-run/remix/assets/6743796/3b4da609-da85-470f-b959-db8a94d25ce0" />
</a>
